simpl/db/mongodb.py

```diff
@@ -0,0 +1,554 @@
+    to always return a write confirmation that was associated with a
+    particular object id, where that object id could be generated
+    *by* pymongo (using ObjectIdInjector) *before* inserting the document.
+
+    This caused a couple problems:
+        1. When performing a partial update of a document,
+           if you did not provide the _id with the data,
+           pymongo would generate a new one, and whatever
+           document you were pointing to (based on a spec)
+           would have its _id overwritten.
+        2. Typically, we would not reach (1), because a
+           partial update will make use of the $set operator,
+           (or any operator) and pymongo would add the _id to
+           that set of data, creating a document that would
+           look something like this:
+                * {'$set': {'data.to.update': 'yes'},
+                   '_id': ObjectId('530264af09df64dd0235b155')}
+           where you would receive the following error:
+
+            pymongo.errors.OperationFailure: Modifiers and non-modifiers
+                                             cannot be mixed
+
+    """
+
+    __collections__ = tuple()
+
+    def __init__(self, connection_string):
+        """Initialize database wrapper."""
+        self.connection_string = connection_string
+        self.safe_connection_string = secrets.hide_url_password(
+            self.connection_string)
+        parsed = pymongo.uri_parser.parse_uri(self.connection_string)
+        self.database_name = parsed['database']
+        self._client = None
+        self._connection = None
+        if eventlet:
+            eventlet.spawn_n(self.tune)
+        else:
+            self.tune()
+
+    @property
+    def client(self):
+        """Return a lazy-instantiated pymongo client."""
+        if eventlet:
+            block = eventlet.semaphore.Semaphore(id(self))
+            with block:
+                if self._client is None:
+                    self._client = pymongo.MongoClient(self.connection_string)
+                    LOG.debug("Created new connection to MongoDB: %s",
+                              self.safe_connection_string)
+        else:
+            if self._client is None:
+                self._client = pymongo.MongoClient(self.connection_string)
+                LOG.debug("Created new connection to MongoDB: %s",
+                          self.safe_connection_string)
+        return self._client
+
+    @property
+    def connection(self):
+        """Connect to and return mongodb database object."""
+        if self._connection is None:
+
+            self._connection = reconnector.MongoProxy(
+                self.client[self.database_name], logger=LOG)
+
+            # pymongo/database.py includes ObjectIdInjector manipulator
+            # by default. Here, we are resetting the manipulators b/c
+            # we don't want the ObjectIdInjector to add an `_id` to
+            # our modifiers.
```

Comment:
The reasoning is pretty detailed in https://github.com/checkmate/simpl/pull/76/files#diff-b304b6c35c39c7c2e29afaf7e6a295c6R116... and still valid, IMO. It's because we wanted to specify our own IDs and support partial updates.